lib/filterx/filterx-metrics-labels.c

```diff
@@ -0,0 +1,361 @@
+  const FilterXMetricsLabel *rhs = *(const FilterXMetricsLabel **) b;
+
+  return strcmp(lhs->name, rhs->name);
+}
+
+static void
+_label_free(FilterXMetricsLabel *self)
+{
+  g_free(self->name);
+  filterx_expr_unref(self->value.expr);
+  g_free(self->value.str);
+
+  g_free(self);
+}
+
+static gchar *
+_init_label_name(FilterXExpr *name)
+{
+  if (!filterx_expr_is_literal(name))
+    {
+      filterx_eval_push_error("failed to initialize metrics label name, name must be a string literal", name, NULL);
+      return NULL;
+    }
+
+  FilterXObject *obj = filterx_expr_eval_typed(name);
+  gchar *str = g_strdup(filterx_string_get_value(obj, NULL));
+  if (!str)
+    filterx_eval_push_error_info("failed to initialize metrics label name, name must be a string literal", name,
+                                 g_strdup_printf("got %s instead", obj->type->name), TRUE);
+
+  filterx_object_unref(obj);
+  return str;
+}
+
+static gboolean
+_init_label_value(FilterXMetricsLabel *self, FilterXExpr *value)
+{
+  if (!filterx_expr_is_literal(value))
+    {
+      self->value.expr = filterx_expr_ref(value);
+      return TRUE;
+    }
+
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+  self->value.str = g_strdup(_format_value_expr(value));
+  scratch_buffers_reclaim_marked(marker);
+
+  return !!self->value.str;
+}
+
+static FilterXMetricsLabel *
+_label_new(FilterXExpr *name, FilterXExpr *value)
+{
+  FilterXMetricsLabel *self = g_new0(FilterXMetricsLabel, 1);
+
+  self->name = _init_label_name(name);
+  if (!self->name)
+    goto error;
+
+  if (!_init_label_value(self, value))
+    goto error;
+
+  return self;
+
+error:
+  _label_free(self);
+  return NULL;
+}
+
+
+struct _FilterXMetricsLabels
+{
+  FilterXExpr *expr;
+  GPtrArray *literal_labels;
+  gboolean is_const;
+};
+
+static void
+_check_label_is_const(gpointer data, gpointer user_data)
+{
+  FilterXMetricsLabel *label = (FilterXMetricsLabel *) data;
+  gboolean *is_const = (gboolean *) user_data;
+
+  if (!(*is_const))
+    return;
+
+  if (!_label_is_const(label))
+    *is_const = FALSE;
+}
+
+gboolean
+filterx_metrics_labels_is_const(FilterXMetricsLabels *self)
+{
+  return self->is_const;
+}
+
+static gboolean
+_format_dict_elem_to_tls_cache(FilterXObject *key, FilterXObject *value, gpointer user_data)
+{
+  const gchar *name_str = _format_str_obj(key);
+  if (!name_str)
+    {
+      filterx_eval_push_error_info("failed to format label name, name must be string", NULL,
+                                   g_strdup_printf("got %s instead", key->type->name), TRUE);
+      return FALSE;
+    }
+
+  const gchar *value_str = _format_str_obj(value);
+  if (!value_str)
+    {
+      filterx_eval_push_error_info("failed to format label value, object does not support string casting", NULL,
+                                   "", FALSE);
+      return FALSE;
+    }
+
+  StatsClusterLabel *label = metrics_cache_alloc_label(metrics_tls_cache());
```

Comment:
actually it might even make sense to cache the metrics_tls_cache() across the entire iteration. Getting the value of a TLS variable is not free, especially when wrapped into an external function, like in this case.

Reply:
Just to clarify, I meant that it would make sense to pass the value around to callbacks that iterate possibly multiple times per message to avoid the TLS overhead.

Reply:
I did not think about the overhead. Thanks, fixed this in a new commit.